File: deepeval/metrics/knowledge_retention/knowledge_retention.py
from typing import Optional, Union, List

from deepeval.test_case import ConversationalTestCase, Turn, TurnParams
from deepeval.metrics import BaseConversationalMetric
from deepeval.metrics.utils import (
    check_conversational_test_case_params,
    construct_verbose_logs,
    trimAndLoadJson,
    initialize_model,
    convert_turn_to_dict,
)
from deepeval.models import DeepEvalBaseLLM
from deepeval.metrics.knowledge_retention.template import (
    KnowledgeRetentionTemplate,
)
from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.metrics.knowledge_retention.schema import (
    Knowledge,
    KnowledgeRetentionVerdict,
    KnowledgeRetentionScoreReason,
)
from deepeval.utils import get_or_create_event_loop, prettify_list


class KnowledgeRetentionMetric(BaseConversationalMetric):
    _required_test_case_params = [TurnParams.CONTENT, TurnParams.ROLE]

    def __init__(
        self,
        threshold: float = 0.5,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        include_reason: bool = True,
        async_mode: bool = True,
        strict_mode: bool = False,
        verbose_mode: bool = False,
    ):
        self.threshold = 1 if strict_mode else threshold
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()
        self.include_reason = include_reason
        self.async_mode = async_mode
        self.strict_mode = strict_mode
        self.verbose_mode = verbose_mode

    def measure(
        self,
        test_case: ConversationalTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ):
        
        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                    )
                )
            else:
                check_conversational_test_case_params(
                    test_case, self._required_test_case_params, self
                )
                
                self.knowledges: List[Union[Knowledge, None]] = (
                    self._generate_knowledges(test_case.turns)
                )
                self.verdicts: List[KnowledgeRetentionVerdict] = (
                    self._generate_verdicts(test_case.turns)
                )
                self.score = self._calculate_score()
                self.reason = self._generate_reason()
                self.success = self.score >= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"Formatted Turns:\n{prettify_list(test_case.turns)}",
                        f"Knowledges:\n{prettify_list(self.knowledges)}",
                        f"Verdicts:\n{prettify_list(self.verdicts)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )
            return self.score

    async def a_measure(
        self,
        test_case: ConversationalTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:
        check_conversational_test_case_params(
            test_case, self._required_test_case_params, self
        )

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self,
            async_mode=True,
            _show_indicator=_show_indicator,
            _in_component=_in_component,
        ):
            self.knowledges: List[Union[Knowledge, None]] = (
                await self._a_generate_knowledges(test_case.turns)
            )
            self.verdicts: List[KnowledgeRetentionVerdict] = (
                await self._a_generate_verdicts(test_case.turns)
            )
            self.score = self._calculate_score()
            self.reason = await self._a_generate_reason()
            self.success = self.score >= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"Knowledges:\n{prettify_list(self.knowledges)}",
                    f"Verdicts:\n{prettify_list(self.verdicts)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )
            return self.score

    async def _a_generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        attritions = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                attritions.append(verdict.reason)

        prompt: dict = KnowledgeRetentionTemplate.generate_reason(
            attritions=attritions,
            score=format(self.score, ".2f"),
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt)
            self.evaluation_cost += cost
            data = trimAndLoadJson(res, self)
            return data["reason"]
        else:
            try:
                res: KnowledgeRetentionScoreReason = (
                    await self.model.a_generate(
                        prompt, schema=KnowledgeRetentionScoreReason
                    )
                )
                return res.reason
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        attritions = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                attritions.append(verdict.reason)

        prompt: dict = KnowledgeRetentionTemplate.generate_reason(
            attritions=attritions,
            score=format(self.score, ".2f"),
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt)
            self.evaluation_cost += cost
            data = trimAndLoadJson(res, self)
            return data["reason"]
        else:
            try:
                res: KnowledgeRetentionScoreReason = self.model.generate(
                    prompt, schema=KnowledgeRetentionScoreReason
                )
                return res.reason
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    async def _a_generate_verdicts(
        self, turns: List[Turn]
    ) -> List[KnowledgeRetentionVerdict]:
        verdicts: List[KnowledgeRetentionVerdict] = []
        for i in range(len(turns)):
            if turns[i].role != "assistant":
                continue

            accumulated_knowledge = [
                knowledge.data
                for knowledge in self.knowledges[:i]
                if knowledge is not None
            ]
            if len(accumulated_knowledge) == 0:
                continue

            prompt = KnowledgeRetentionTemplate.generate_verdict(
                llm_message=turns[i].content,
                accumulated_knowledge=accumulated_knowledge,
            )
            if self.using_native_model:
                res, cost = await self.model.a_generate(prompt)
                self.evaluation_cost += cost
                data = trimAndLoadJson(res, self)
                verdict = KnowledgeRetentionVerdict(**data)
            else:
                try:
                    verdict: KnowledgeRetentionVerdict = (
                        await self.model.a_generate(
                            prompt, schema=KnowledgeRetentionVerdict
                        )
                    )
                except TypeError:
                    res = await self.model.a_generate(prompt)
                    data = trimAndLoadJson(res, self)
                    verdict = KnowledgeRetentionVerdict(**data)
            verdicts.append(verdict)
        return verdicts

    def _generate_verdicts(
        self, turns: List[Turn]
    ) -> List[KnowledgeRetentionVerdict]:
        verdicts: List[KnowledgeRetentionVerdict] = []
        for i in range(len(turns)):
            if turns[i].role != "assistant":
                continue

            accumulated_knowledge = [
                knowledge.data
                for knowledge in self.knowledges[:i]
                if knowledge is not None
            ]
            if len(accumulated_knowledge) == 0:
                continue

            prompt = KnowledgeRetentionTemplate.generate_verdict(
                llm_message=turns[i].content,
                accumulated_knowledge=accumulated_knowledge,
            )

            if self.using_native_model:
                res, cost = self.model.generate(prompt)
                self.evaluation_cost += cost
                data = trimAndLoadJson(res, self)
                verdict = KnowledgeRetentionVerdict(**data)
            else:
                try:
                    verdict: KnowledgeRetentionVerdict = self.model.generate(
                        prompt, schema=KnowledgeRetentionVerdict
                    )
                except TypeError:
                    res = self.model.generate(prompt)
                    data = trimAndLoadJson(res, self)
                    verdict = KnowledgeRetentionVerdict(**data)
            verdicts.append(verdict)
        return verdicts

    async def _a_generate_knowledges(
        self, turns: List[Turn]
    ) -> List[Union[Knowledge, None]]:
        knowledges: List[Union[Knowledge, None]] = [None] * len(turns)

        for i in range(0, len(turns)):
            if turns[i].role == "assistant":
                continue

            previous_turns = turns[:i]
            user_message = turns[i].content

            prompt = KnowledgeRetentionTemplate.extract_data(
                user_message=user_message,
                previous_turns=[
                    convert_turn_to_dict(turn) for turn in previous_turns
                ],
            )
            if self.using_native_model:
                res, cost = await self.model.a_generate(prompt)
                self.evaluation_cost += cost
                data = trimAndLoadJson(res, self)
                knowledges[i] = Knowledge(data=data)
            else:
                try:
                    knowledges[i] = await self.model.a_generate(
                        prompt, schema=Knowledge
                    )
                except TypeError:
                    res = await self.model.a_generate(prompt)
                    data = trimAndLoadJson(res, self)
                    knowledges[i] = Knowledge(data=data)

        return knowledges

    def _generate_knowledges(
        self, turns: List[Turn]
    ) -> List[Union[Knowledge, None]]:
        knowledges: List[Union[Knowledge, None]] = [None] * len(turns)

        for i in range(0, len(turns)):
            if turns[i].role == "assistant":
                continue

            previous_turns = turns[:i]
            user_message = turns[i].content

            prompt = KnowledgeRetentionTemplate.extract_data(
                user_message=user_message,
                previous_turns=[
                    convert_turn_to_dict(turn) for turn in previous_turns
                ],
            )

            if self.using_native_model:
                res, cost = self.model.generate(prompt)
                self.evaluation_cost += cost
                data = trimAndLoadJson(res, self)
                knowledges[i] = Knowledge(data=data)
            else:
                try:
                    knowledges[i] = self.model.generate(
                        prompt, schema=Knowledge
                    )
                except TypeError:
                    res = self.model.generate(prompt)
                    data = trimAndLoadJson(res, self)
                    knowledges[i] = Knowledge(data=data)

        return knowledges

    def _calculate_score(self) -> float:
        number_of_verdicts = len(self.verdicts)
        if number_of_verdicts == 0:
            return 0

        retention_count = 0
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "no":
                retention_count += 1

        score = retention_count / number_of_verdicts

        return 0 if self.strict_mode and score < self.threshold else score

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.score >= self.threshold
            except:
                self.success = False
        return self.success

    @property
    def __name__(self):
        return "Knowledge Retention"

File: typescript/src/metrics/knowledge-retention/knowledge-retention.ts
import { BaseConversationalMetric } from "@/metrics/base-conversational-metric";
import { resolveThreshold } from "@/metrics/base-metrics";
import { ConversationalTestCase, MultiTurnParams, Turn } from "@/test-case";
import { DeepEvalBaseLLM } from "@/models";
import type { DeepEvalBaseSystemOneModel } from "@/models/system-one";
import type { EvalModeName } from "@/config/eval-mode";
import {
  initializeMetricModels,
  generateWithSchema,
  constructVerboseLogs,
  prettifyList,
} from "@/metrics/utils";
import {
  parseQuestions,
  runSystemOneEval,
  systemOneProbability,
  verdictFromProbability,
  type SystemOneBinarySpec,
  type SystemOneEvalSpec,
} from "@/metrics/system-one";
import {
  checkConversationalTestCaseParams,
  convertTurnToDict,
} from "@/metrics/conversational-utils";
import {
  KnowledgeSchema,
  KnowledgeRetentionVerdictSchema,
  KnowledgeRetentionScoreReasonSchema,
  type Knowledge,
  type KnowledgeRetentionVerdict,
} from "@/metrics/knowledge-retention/schema";
import { type MetricTemplateOverride } from "@/templates/override";

const TEMPLATE_CLASS = "KnowledgeRetentionMetric";

export type KnowledgeRetentionTemplateOverride =
  MetricTemplateOverride<"KnowledgeRetentionMetric">;

export interface KnowledgeRetentionMetricOptions {
  threshold?: number | null;
  flaky?: boolean;
  model?: DeepEvalBaseLLM | string;
  /** The System One model (Jev) used under `hybrid` / `system_one`. */
  systemOneModel?: DeepEvalBaseSystemOneModel | string;
  /** Who decides; defaults to `DEEPEVAL_EVAL_MODE`, then `llm`. */
  evalMode?: EvalModeName;
  includeReason?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  evaluationTemplate?: KnowledgeRetentionTemplateOverride;
}

/**
 * Knowledge Retention — does the assistant remember facts the user established
 * earlier (no attrition)? Extract knowledge per user turn, then judge each
 * assistant turn against the accumulated knowledge. Score = retained / total.
 * **Higher is better** (`success = score >= threshold`).
 */
export class KnowledgeRetentionMetric extends BaseConversationalMetric {
  knowledges: (Knowledge | null)[] = [];
  verdicts: KnowledgeRetentionVerdict[] = [];

  constructor(options: KnowledgeRetentionMetricOptions = {}) {
    const strictMode = options.strictMode ?? false;
    super(strictMode ? 1 : resolveThreshold(options.threshold, 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
      flaky: options.flaky,
      evaluationTemplate: options.evaluationTemplate,
    });
    this.templateClass = TEMPLATE_CLASS;
    this.requiredParams = [MultiTurnParams.CONTENT, MultiTurnParams.ROLE];
    initializeMetricModels(this, options);
  }

  async measure(testCase: ConversationalTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      checkConversationalTestCaseParams(testCase, this.requiredParams, this);
      this.evaluationCost = this.usingNativeModel ? 0 : undefined;
      if (await runSystemOneEval(this, testCase)) return this.score as number;

      this.knowledges = await this.generateKnowledges(testCase.turns);
      this.verdicts = await this.generateVerdicts(
        testCase.turns,
        testCase.multimodal,
      );
      this.score = this.calculateScore();
      this.reason = await this.generateReason();
      this.success = this.isSuccessful();

      this.verboseLogs = constructVerboseLogs(this, [
        `Knowledges:\n${prettifyList(this.knowledges)}`,
        `Verdicts:\n${prettifyList(this.verdicts)}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  /** Extract knowledge from each user turn (assistant turns get `null`). */
  private async generateKnowledges(
    turns: Turn[],
  ): Promise<(Knowledge | null)[]> {
    const knowledges: (Knowledge | null)[] = new Array(turns.length).fill(null);
    const extracted = await Promise.all(
      turns.map(async (turn, i) => {
        if (turn.role === "assistant") return null;
        const prompt = this.getPrompt("extract_data", {
          user_message: turn.content,
          previous_turns: turns.slice(0, i).map((t) => convertTurnToDict(t)),
        });
        return generateWithSchema(this, prompt, KnowledgeSchema);
      }),
    );
    extracted.forEach((k, i) => {
      if (k != null) knowledges[i] = k;
    });
    return knowledges;
  }

  /** One verdict per assistant turn that has prior accumulated knowledge. */
  private async generateVerdicts(
    turns: Turn[],
    multimodal: boolean,
  ): Promise<KnowledgeRetentionVerdict[]> {
    const results = await Promise.all(
      turns.map(async (turn, i) => {
        if (turn.role !== "assistant") return null;
        const accumulatedKnowledge = this.knowledges
          .slice(0, i)
          .filter((k): k is Knowledge => k != null && k.data != null)
          .map((k) => k.data);
        if (accumulatedKnowledge.length === 0) return null;
        const p = await systemOneProbability(
          this,
          this.systemOneVerdictSpec(
            turn.content,
            accumulatedKnowledge,
            multimodal,
          ),
        );
        if (p !== undefined) {
          return {
            verdict: verdictFromProbability(p),
            reason: `P(yes)=${p.toFixed(2)}`,
          };
        }
        const prompt = this.getPrompt("generate_verdict", {
          llm_message: turn.content,
          accumulated_knowledge: accumulatedKnowledge,
        });
        return generateWithSchema(
          this,
          prompt,
          KnowledgeRetentionVerdictSchema,
        );
      }),
    );
    return results.filter((v): v is KnowledgeRetentionVerdict => v != null);
  }

  private systemOneVerdictSpec(
    llmMessage: string,
    accumulatedKnowledge: unknown[],
    multimodal: boolean,
  ): SystemOneBinarySpec | undefined {
    if (multimodal) return undefined;
    return {
      instructions: this.getPrompt("_experimental_system_one_verdict"),
      state: {
        llm_message: llmMessage,
        accumulated_knowledge: accumulatedKnowledge,
      },
    };
  }

  systemOneEvalSpec(
    testCase: ConversationalTestCase,
  ): SystemOneEvalSpec | undefined {
    if (testCase.multimodal) return undefined;
    return {
      evaluationParams: this.requiredParams,
      questions: parseQuestions(
        this.getPrompt("_experimental_system_one_questions"),
      ),
    };
  }

  private async generateReason(): Promise<string | undefined> {
    if (!this.includeReason) return undefined;
    const attritions = this.verdicts
      .filter((v) => v.verdict.trim().toLowerCase() === "yes")
      .map((v) => v.reason);
    const prompt = this.getPrompt("generate_reason", {
      attritions,
      score: (this.score ?? 0).toFixed(2),
    });
    const { reason } = await generateWithSchema(
      this,
      prompt,
      KnowledgeRetentionScoreReasonSchema,
    );
    return reason;
  }

  /** Score = fraction of assistant turns with NO attrition ("no" verdicts). */
  private calculateScore(): number {
    const total = this.verdicts.length;
    if (total === 0) return 1;
    const retained = this.verdicts.filter(
      (v) => v.verdict.trim().toLowerCase() === "no",
    ).length;
    const score = retained / total;
    return this.applyStrictMode(score);
  }

  get name(): string {
    return "Knowledge Retention";
  }
}
